packages/solid/store/src/store.ts

```diff
@@ -223,11 +223,11 @@ export type DeepReadonly<T> = {
 export type StoreSetter<T> =
   | T
   | Partial<T>
-  | ((prevState: T, traversed?: (keyof any)[]) => Partial<T> | void);
+  | ((prevState: DeepReadonly<T>, traversed?: (keyof any)[]) => Partial<DeepReadonly<T>> | void);
```

Comment:
I haven't gotten to test inference here yet. But if there's a way to fix the broken inference you mentioned, while keeping the "correct" readonly types, that'd be nice, so as to prevent people doing the wrong thing.

Reply:
This PR changes two things:

* The types of `SetStoreFunction` and `StoreSetter` such that `DeepReadonly` isn't applied directly at the root of the former, and instead only where the user would "interact" with the store in the latter. 
* `produce` now accepts all types that are part of the union of `StoreSetter` so that it can use the `T` type included in that type to properly type its own parameters, avoiding the inference issue. 

Do you mean that `StoreSetter` should also accept `DeepReadonly` variants of `T | Partial<T>`?